### DISCUSSION TITLE
Mode switches (a suggestion for a new feature)

### DISCUSSION THREAD

**Discussion by kjoetom** [8VIM/8VIM#186] [Ideas]
Hi guys,

I've been thinking for a while about how to use the layout of 8vim for additional operations besides just typing letters.

And I had the idea to introduce something like "mode switches". By this I mean gestures that switch 8vim from the "standard mode" to other "special modes", in order to be able to enter diacritical characters, special characters, numbers, smileys and punctuations.

That would be quite in the sense of vim, wouldn't it? ;-)

As a gesture to switch to a special mode, I could imagine moving the finger from the center to one of the quadrants and then immediately back again:

![1](https://user-images.githubusercontent.com/77550697/119580434-ac7a5180-bdc0-11eb-9079-7da451f19c7b.png)

This way one could select - depending on the quadrant - 4 different modes.
And, depending on how many times one performs this gesture in a row, one could also achieve different functions.

See the following example to get a better understanding of what I mean.

![2](https://user-images.githubusercontent.com/77550697/119580679-27dc0300-bdc1-11eb-8c0f-4287ca59d223.png)

Let me illustrate below how I envision this using the example of diacritical letters.

The scenarios that could be implemented in a "diacritics character mode" would be the following:

1. writing a diacritic letter, either ...
- a. ... one out of 7 predefined main diacritic letters (the most common ones you need), or
- b. ... a predefined one based on a selectable basic letter, or
- c. ... any one based on a selectable basic letter

2. programming of the predefined diacritic letters, either ...
- a. ... one of the 7 predefined main diacritic letters, or
- b. ... those that are predefined when selecting a particular basic letter

![3](https://user-images.githubusercontent.com/77550697/119580915-b94b7500-bdc1-11eb-8655-8df93b1f0225.png)

![4](https://user-images.githubusercontent.com/77550697/119580919-bbadcf00-bdc1-11eb-9ca0-5a4396f97efe.png)

![5](https://user-images.githubusercontent.com/77550697/119580923-bea8bf80-bdc1-11eb-8936-80248870dbf9.png)

![6](https://user-images.githubusercontent.com/77550697/119581212-6920e280-bdc2-11eb-9cb7-c09d5ef57f7b.png)

![7](https://user-images.githubusercontent.com/77550697/119580931-c36d7380-bdc1-11eb-8137-e1834dbb3d55.png)

I know this is kind of complex and maybe not immediately understandable.
I wanted to introduce the idea to you anyway.

In a similar way as the "diacritic character mode" you could also build the other special modes.

Of course, I have no idea if this is even feasible from a programming point of view.

What do you think of the idea?

**Comment by nawordar** [comment]
This looks great! I believe that for Polish layout layers/modes are a necessity. It has too many diactrics to fit on a single layer.

**Comment by kjoetom** [comment]
Hello and all the best for 2022, which brought us iteration 8 of 8vim. I'm glad to see that it is still being developed.

I would like to ask again if it would be possible to create a universal input variant of diacritical characters after all (a simpler version of what I've described above).

If it would be possible to implement such a switch, which would add diacritical marks to the last written letter, 8vim would suddenly be usable for many languages (independent of the keyboard layout used).

For example, to write an "ä", one would have to do the following:

![8vim-xc](https://user-images.githubusercontent.com/77550697/148314632-b650988c-01fa-4232-869c-30388dcd32ea.png)

The question is whether you think this is a good idea and whether such a function could be implemented easily.

**Comment by flide** [comment]
@kjoetom, I don't really know how I could have missed on your post. Sorry for the 'delay', but here are my thoughts.

Although the idea looks great on the surface but uses up too many gestures, most of which I already plan to use for selecting candidates for word predictions. Having said that, your post did inspire an idea.

We will use the following stroke for this idea. Once you type a letter, this particular gesture will start cycling through the letter's diacritic variants. At first, this might sound tedious, but if we complement it with adaptive cycling - meaning, more used variants pop early - then within a few days the most commonly used diacritic variant would be at the top of the stack. This should be generic and adaptable enough to fit everyone's needs without the need to setup and remember a new X-board.

![8vim_down and back stroke](https://user-images.githubusercontent.com/7330631/148420535-135785ca-4953-46a8-aca1-00aad3aeeb12.png)

**Reply by kjoetom** [reply]
@flide, thank you for your quick feedback.
I find your suggestion with the adaptive cycling diacritic variants really inspiring. This way it is even much easier to set up your own preferred diacritic letters than I wanted to implement in my thoughts in the post at the very top.

Two more aspects come to my mind in this context, which I would like to mention here briefly.

Perhaps in this way not only diacritical letters could be realized, but at the same time also "related letters" e.g. 
- the letters of the Greek alphabet, like "α" being related to "a", or even 
- cyrilic letters, like "с" being related to "s" or 
- ligated letters like "æ" being related to "a", or
- special characters like the german "ß" being related to "s"

To allow for continuity of a personalized layout, it might be helpful to be able to stop (turn off) the adaptive process at a certain point.

**Comment by lguarda** [comment]
Hello,
First thanks for this keyboard i found the idea really nice.

I'm currently using the keyboard but im pretty sure i will never be able to beat the speed of my two thumbs
However since i'm practicing almost every day, i noticed (pretty sure everyone noticed it) that every key after the first 'row'
slowed me down a lot and of course it worst for every 'row' since it's increase the travel distance, again i don't think anybody learn something new here
But it's make me thinks of something, we could shrink a little bit more the max travel of up to 4 more letter using the same principle used in this discussion

I'm curious of what people think of this but having 8 letter instead of 4 on a very very short travel distance seems appealing
and it will also shorten the distance of all other key by one 'row'

![8vim_faster](https://github.com/user-attachments/assets/161e4e03-4a38-4115-86b9-2e2bf5717119)

What do you thinks of this?
